torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,395 @@
+            val,
+            " and ",
+            *rhs);
+      }
+    }
+  }
+
+  // Prepraring for clean up phase. At this point, record all  subModules that
+  // contains mutable attributes.
+  void recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* subBlock : n->blocks()) {
+          blocks.push(subBlock);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto& name = n->s(attr::name);
+          if (module_.hasattr(name)) {
+            auto attr = module_.attr(name);
+            insertMutableAttr(name, attr, module_);
+          }
+        }
+      }
+    }
+  }
+
+  IValue overrideGradient(IValue attr) {
+    if (attr.isTensor()) {
+      auto t = attr.toTensor();
+      if (t.requires_grad()) {
+        t = autograd::as_variable_ref(t).detach();
+        t.set_requires_grad(false);
+        attr = IValue(t);
+      }
+    } else if (attr.isTuple()) {
+      std::vector<IValue>& elems = attr.toTuple()->elements();
+      for (auto& elem : elems) {
+        elem = overrideGradient(elem);
+      }
+    } else if (attr.isList()) {
+      c10::List<IValue> elems = std::move(attr).toList();
+      for (size_t i = 0; i < elems.size(); i++) {
+        elems.set(i, overrideGradient(elems.extract(i)));
+      }
+      attr = std::move(elems);
+    }
+
+    return attr;
+  }
+
+  void propagateAttributes(std::shared_ptr<Graph>& graph) {
+    std::unordered_map<
+        script::ModulePtr,
+        std::unordered_map<std::string, Value*>>
+        attrValues;
+    auto isEval = !module_.is_training();
+    GRAPH_DEBUG("Freezing Module in ", isEval ? "eval mode" : "training mode");
+    auto block = graph->block();
+    std::stack<Block*> blocks({block});
+
+    // Record Attributes that are explicitely set in the module. They cannot be
+    // folded.
+    recordMutableAttrs(graph);
+
+    Node* m = *block->nodes().begin();
+    WithInsertPoint guard(m);
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+        Node* n = *it;
+        it++; // advance iterator bc the current node may be destroyed
+
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          auto inputNode = n->inputs()[0]->node();
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            GRAPH_DEBUG("attribute: ", name, " is mutable.")
+            continue;
+          }
+          TORCH_INTERNAL_ASSERT(attrModule.hasattr(name));
+          Value* paramConst = nullptr;
+          auto I = attrValues.find(attrModule._ivalue());
+          if (I != attrValues.end()) {
```

Comment:
nit: the `I` / `II` variable names are confusing to read.

Reply:
Agreed, sorry in LLVM we capitalize variables. I am renaming I to iter and II to iter2